stdlib/src/collections/string.mojo

```diff
@@ -686,6 +686,63 @@ fn isprintable(c: UInt8) -> Bool:
     return ord_space <= int(c) <= ord_tilde
 
 
+# ===----------------------------------------------------------------------=== #
+# String Representation
+# ===----------------------------------------------------------------------=== #
+
+
+struct StringRepr(Writable):
+    """Represents a string representation of an String."""
+
+    var _value: StringSlice[ImmutableAnyOrigin]
+    """The string representation of the String."""
+
+    fn __init__(inout self, value: String):
+        """Constructs a string representation of the string.
+
+        Args:
+            value: The string to represent.
+        """
+        self._value = StringSlice[ImmutableAnyOrigin](
+            unsafe_from_utf8_ptr=value.unsafe_ptr(),
+            len=len(value),
+        )
+
+    fn write_to[W: Writer](self, inout writer: W):
+        """Writes the string representation to the provided writer.
+
+        Parameters:
+            W: The type of the writer.
+
+        Args:
+            writer: The writer to write the string representation to.
+        """
+        var quote = "'"
+        if len(self._value) > 0 and self._value[0] == "'":
+            quote = '"'
+        writer.write(quote)
+        for s in self._value:
+            if s == "\\":
+                writer.write(r"\\")
+            elif s == "\t":
+                writer.write(r"\t")
+            elif s == "\n":
+                writer.write(r"\n")
+            elif s == "\r":
+                writer.write(r"\r")
+            else:
+                var codepoint = ord(s)
+                if isprintable(codepoint):
+                    writer.write(s)
+                elif codepoint < 0x10:
+                    writer.write(hex(codepoint, prefix=r"\x0"))
+                elif codepoint < 0x20 or codepoint == 0x7F:
+                    writer.write(hex(codepoint, prefix=r"\x"))
+                else:  # multi-byte character
+                    writer.write(s)
+        writer.write(quote)
```

Comment:
This isn't really much faster. It still uses write `hex()` and doesn't alloc before doing everything

Reply:
Yes, it was not a great optimization. I only mapped the top-level constructor to use `.write()` but did not spend the needed work on making it way faster.